File: revisiting_ctci/StringCompression_1_6.go

/*
take a string and compress it into repeated digit counts:

AAAATTTTTGGG
A4T5G3

if the new string is longer than the original, then return the original.

*/

package main

import (
	"fmt"
)

func CompressString(s string) string {
	outstring := ""

	var current_letter rune
	var current_count int

	for _, char := range s {
		if char == current_letter{
			current_count += 1
		}else{
			if current_count != 0 {
				addition := fmt.Sprintf("%v%v", string(current_letter), current_count)
				outstring += addition
			}
			current_letter = char
			current_count = 1

		}
	}
	addition := fmt.Sprintf("%v%v", string(current_letter), current_count)
	outstring += addition

	if len(outstring) < len(s){
		return outstring
	}
	return s
}


func main() {

	t1 := "AAAATTTTTGGG"
	o1 := "A4T5G3"

	fmt.Println(t1)
	fmt.Println(CompressString(t1))
	fmt.Println(o1)
	fmt.Println(CompressString(t1) == o1)

	t2 := "ATGCATGCATGC"
	o2 := "ATGCATGCATGC"

	fmt.Println(t2)
	fmt.Println(o2)
	fmt.Println(CompressString(t2) == o2)

	t3 := "CGGGAAAATTTTTGGG"
	o3 := "C1G3A4T5G3"

	fmt.Println(t3)
	fmt.Println(CompressString(t3))
	fmt.Println(o3)
	fmt.Println(CompressString(t3) == o3)

}

File: revisiting_ctci/StringCompression_1_6.py
"""
take a string and compress it into repeated digit counts:

AAAATTTTTGGG
A4T5G3

if the new string is longer than the original, then return the original.

"""

import unittest


def StringCompression(s):
	outstring = ''
	current_char = ''
	current_count = 0

	for i in s:
		if i == current_char:
			current_count += 1

		else:
			if current_count != 0:
				outstring+= current_char
				outstring+=str(current_count)

			current_char = i
			current_count = 1

	outstring+= current_char
	outstring+=str(current_count)

	if len(outstring) < len(s):
		return outstring
	return s


class TestStringCompression(unittest.TestCase):
	def test_StringCompression(self):
		t1 = "AAAATTTTTGGG"
		o1 = "A4T5G3"
		t2 = "ATGCATGCATGC"
		o2 = "ATGCATGCATGC"
		t3 = "CGGGAAAATTTTTGGG"
		o3 = "C1G3A4T5G3"

		self.assertEqual(StringCompression(t1), o1)
		self.assertEqual(StringCompression(t2), o2)
		self.assertEqual(StringCompression(t3), o3)

if __name__ == '__main__':
	unittest.main()